applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp

```diff
@@ -79,6 +86,154 @@ class GeoTransportEquationUtilities
+
+        for (unsigned int g_point = 0; g_point < number_G_points; ++g_point) {
+            const double degree_of_saturation =
+                CalculateDegreeOfSaturation(row(N_container, g_point), pressure_vector,
+                                            retention_parameters, rRetentionLawVector[g_point]);
+
+            const double density = CalculateSoilDensity(degree_of_saturation, rProp);
+
+            GeoElementUtilities::AssembleDensityMatrix(density_matrix, density);
+            GeoElementUtilities::CalculateNuMatrix<TDim, TNumNodes>(Nu, N_container, g_point);
+            noalias(aux_density_matrix) = prod(density_matrix, Nu);
+
+            // Adding contribution to Mass matrix
+            const double integration_coefficient_initial_configuration =
+                CalculateIntegrationCoefficientInitialConfiguration(
+                    g_point, rGeom, integration_points, IntegrationMethod, rStressStatePolicy);
+            GeoElementUtilities::AssembleUUBlockMatrix(
+                mass_matrix, prod(trans(Nu), aux_density_matrix) * integration_coefficient_initial_configuration);
+        }
+        return mass_matrix;
+    }
+
+private:
+    static double CalculateIntegrationCoefficientInitialConfiguration(
+        unsigned int                                      GPoint,
+        const Geometry<Node>&                             rGeom,
+        const Geometry<Node>::IntegrationPointsArrayType& rIntegrationPoints,
+        const GeometryData::IntegrationMethod             IntegrationMethod,
+        const StressStatePolicy&                          rStressStatePolicy)
+    {
+        Matrix J0;
+        Matrix inv_J0;
+        GeometryUtils::JacobianOnInitialConfiguration(rGeom, rIntegrationPoints[GPoint], J0);
+        const Matrix& dN_De = rGeom.ShapeFunctionsLocalGradients(IntegrationMethod)[GPoint];
+        double        det_J_initial_configuration;
+        MathUtils<double>::InvertMatrix(J0, inv_J0, det_J_initial_configuration);
+        Matrix dNu_dX_initial_configuration;
+        GeometryUtils::ShapeFunctionsGradients(dN_De, inv_J0, dNu_dX_initial_configuration);
+
+        return rStressStatePolicy.CalculateIntegrationCoefficient(
+            rIntegrationPoints[GPoint], det_J_initial_configuration, rGeom);
+    }
+
+    static double CalculateDegreeOfSaturation(const Vector&                rNVector,
+                                              const Vector&                rPressureVector,
+                                              RetentionLaw::Parameters&    rRetentionParameters,
+                                              const RetentionLaw::Pointer& rRetentionLaw)
+    {
+        const double fluid_pressure =
+            GeoTransportEquationUtilities::CalculateFluidPressure(rNVector, rPressureVector);
+        rRetentionParameters.SetFluidPressure(fluid_pressure);
+
+        return rRetentionLaw->CalculateSaturation(rRetentionParameters);
+    }
```

Comment:
In principle yes; however, it will be useful to talk to @avdg81  and @rfaasse because I see some difficulties to move this function to a retention law class.